### ISSUE TITLE
Parachains low-level networking meta-issue

### ISSUE THREAD

**Issue by tomaka** [paritytech/polkadot-sdk#989]
I'm opening this as a meta-issue of things to do on the low-level networking side of things in order to make parachains work properly. This issue is meant to be updated over time with the things that remain to do. **If something isn't implemented yet and is not in this list, then you should say something, as otherwise it might not get any work.**

"Low-level networking" is defined as the APIs and utilities on top of which the actual high-level protocols (such as collation and PoV distribution) are implemented. This encompasses gossiping, sending directed messages, and connecting to the nodes we want to be connected to.

## Introduction

Here is a small introduction and definition of terms:

- **Relay chain**: Since the word "Polkadot" can mean a lot of different things, we use the word "relay chain" to designate the main Polkadot chain. https://telemetry.polkadot.io/#/Polkadot
- **Validator**: An authority node of the relay chain. The list of validators is rotated every 4 hours. A validator is identified by a public key which is not the same as the `PeerId`.
- **Authority-discovery**: The `sc-authority-discovery` crate in Substrate. Using this code, validators publish on the DHT records whose key is their validator public key, and whose content is their `PeerId`, their multiaddresses, and a signature. This makes it possible for everyone to find out how to connect to a certain validator.
- **Sentry node**: Full node whose aim is to protect a validator node. The underlying idea is: one shouldn't be able to directly connect to a validator, but rather all of a validator's comunications should go through its sentry node(s). Sentry nodes are deprecated and support will eventually be removed https://github.com/paritytech/substrate/issues/6845.
- **Parachain**: Designates a blockchain that is separate from the relay chain but "connected to" Polkadot. (This is the entire point of Polkadot and I invite you to read more if you don't know what a parachain is).
- **Collating**: Sending a block from a parachain to a validator for that validator to verify it.
- **Collator**: Any node on a parachain that is collating. There isn't any specific authorization required to be a collator. Any node can do it.

The network protocols that the Polkadot binary needs to support can be divided into three categories:

- Regular relay chain communication: block announces, block requests, transactions, grandpa, and minor other things. This is already implemented and well-working and is included here for the sake of completeness.
- Communication between validators only. The exact protocols here are still to be precisely defined, but will consist in a combination of gossiping, directed messages, and pull-based protocols.
- Communication between collators and validators. Any node, even not connected to the relay chain, should be able to connect to validators and send a block. In other words, validators can be seen as public endpoints that anyone can use to propose a block from a parachain.

Practically speaking, parachain implementations will be based on Substrate and therefore be libp2p networks themselves. TCP/QUIC connections can thus be shared between the parachains and relay chain.
Theoretically, though, they don't *have to*. Validators don't know any of the technical details of parachains. All that validators should do is provide a libp2p protocol that lets anyone who connects propose a parachain block.

The Polkadot repository contains both the code that collators would run to submit blocks (including the networking), and also the code that validators run to verify these block proposals and communicate with each other.

## Areas of work

**The three things we need to work on are:**

- Making sure that collators properly connect to the validators assigned to their parachain.
- Making sure that validators cannot get their resources exhausted by collators (i.e. preventing DoS attacks). This is necessary because collators aren't authenticated, and is where https://github.com/paritytech/polkadot/issues/1348 comes into play. There is no miracle solution for preventing DoS attacks, and we'll go for mitigation techniques.
- Providing code that the rest of the parachain implementation team can leverage to write the high-level protocols, most notably the intra-validators protocols.

All the sections below describe work that belongs to one of these three categories.

## Discovery and connectivity

Collators should (but don't strictly need to) be connected to the relay chain just like any relay chain node would be. In other words, should be connected to a random set of full nodes and validators.
But they should also be connected to a random set of the validators assigned to their parachain. The interval at which the list of validators assigned to a parachain rotates is still to be defined, but should be around one to five minutes.

Collators determine the validators assigned to their parachain through the relay chain runtime, although that is out of scope of this issue. Once they have a list of validator identities, collators use the `authority-discovery` module to determine the corresponding `PeerId` and multiaddresses.

What happens afterwards is still to be clarified. The straight-forward idea would be to put them in a "priority group" of the peerset.

- [x] https://github.com/paritytech/substrate/issues/6595
- [x] https://github.com/paritytech/substrate/issues/6087
- [x] https://github.com/paritytech/substrate/issues/6615
- [x] https://github.com/paritytech/polkadot/issues/1392
- [x] https://github.com/paritytech/substrate/issues/6264
- [x] https://github.com/paritytech/substrate/issues/6772
- [x] https://github.com/paritytech/polkadot/issues/1544 - Re-enable authority-discovery again
- [x] Write the Polkadot code that plugs this all together, so that we do actually connect to the parachain-assigned validators.
- [x] https://github.com/paritytech/substrate/issues/6910
- [ ] https://github.com/paritytech/substrate/issues/7072
- [x] https://github.com/paritytech/substrate/issues/7461

## Providing helpers for request-responses

First of all, some context: the difference between notifications and request-responses is that request-responses have their timeout automatically tracked, are processed in parallel (you continue to receive notifications while the request is being processed), can have higher message size limits, are tracked separately on Prometheus/Grafana, don't risk being interrupted by a rotation of the peer connections, and so on. The request answering code (as in, answering requests made by other peers) can also be done in parallel, and assigned a lower CPU/IO priority. I do believe that request-responses should be their own thing and not something implemented on top of notifications.

- [x] https://github.com/paritytech/substrate/issues/5536
- [x] Request-reponse protocols need to directly reach their target, and sentry nodes are an obstacle: https://github.com/paritytech/substrate/issues/6762
- [ ] Request-responses should make it possible to prioritize some peers' requests over others.
- [ ] Plug this into Polkadot.

## Properly handling back-pressure when sending notifications

Considering that validators might want to apply back-pressure on collators, it becomes important that collators respond well to this back-pressure and don't just have a buffer of messages whose size keeps increasing. This change isn't specific to collators, though, and all of the Polkadot networking should benefit from proper back-pressure.

- [x] First, do https://github.com/paritytech/substrate/issues/6402 in order to tackle https://github.com/paritytech/substrate/issues/5481.
- [ ] Add some helper that makes it easy to use this mechanism: https://github.com/paritytech/polkadot/issues/1453.

## Properly implementing back-pressure when receiving notifications

In order to mitigate the risk for DoS attacks, we notably want to be able to pull messages from more trusted peers at a higher rate than others.

At the moment, `sc-network` exposes an API that lets you receive notifications one by one. You pull a notification, then the next, then the next, and so on.
If we do want to prioritize certain peers, then we need to pull notifications from some peers more often than others:

- [ ] https://github.com/paritytech/polkadot-sdk/issues/554
- [ ] As it would be quite complicated to directly use an API that lets you pull notifications from individual peers, we should add in Substrate some code that wraps around this mechanism and still lets you process notifications in a linear way (one by one), while receiving feedback about the priority to assign to each peer.
- [ ] Put it in Polkadot.

## Misc

Not critical issues, but would be nice to have:

- [ ] https://github.com/paritytech/substrate/issues/6605
- [ ] Exchanging whitelists and blacklists of collating peers between validators. I'm putting this in "Misc" because it is at the intersection of many different areas, and it's completely unclear how we're going to implement that.
- [x] https://github.com/paritytech/substrate/issues/6241, otherwise spurious warnings get printed all the time
- [x] https://github.com/paritytech/substrate/issues/6402, for performances
- [x] https://github.com/paritytech/substrate/issues/5825, while not strictly related to Polkadot, would greatly improve the general networking experience.
- [x] https://github.com/libp2p/rust-libp2p/issues/211

**Comment by Hugo-Trentesaux** [comment]
> That's actually on the way, but we need to remove the so-called legacy substream first, because it is invasive and is incompatible with that idea (https://github.com/paritytech/substrate/issues/5670). 

@tomaka, what is the status of the legacy substream removal? Issue https://github.com/paritytech/substrate/issues/5670 is closed as completed but there is still some comments in the code pointing to closed issues related to this. Like:

https://github.com/paritytech/polkadot-sdk/blob/21930ed2019219c2ffd57c08c0bf675db467a91f/substrate/client/network/src/protocol/message.rs#L119-L124
